### PR TITLE
[bndtools test] Wait for cnf workspace before proceeding

### PR DESCRIPTION
**Pull request by kriegfrj**
In order to try and overcome race conditions at startup, put in a callback that causes the tester to wait for the Workspace to be ready before proceeding.

This is a bit of a hack, but it seems to stabilise the tests - I ran it in my own CI instance on GitHub a few times and it seems to have succeeded on all platforms each time.

### CODE REVIEW

**Comment by kriegfrj** [comment]
Looking at the failure, it seems unrelated to the previous. It seems that BuildPathQuickFixProcessorTest_WithSimpleOnClasspath failed to add Bndtools.core.test.fodder.simple to the -buildpath. I've never seen this before and when I reran it it passed. 

Perhaps when it attempted to install the simple bundle onto -buildpath, the local repository hadn't finished updating yet?

File: bndtools.core/src/bndtools/central/Central.java [hunk 1]
```diff
@@ -240,9 +240,10 @@ public static Workspace getWorkspace() throws Exception {
 				// one
 				File workspaceDirectory = getWorkspaceDirectory();
 				// Check to see if we need to convert it...
-				if (workspaceDirectory != null && ws.isDefaultWorkspace()) {
```
Comment (bjhargrave) [inline bndtools.core/src/bndtools/central/Central.java:243]: You should pull this commit out into its own PR so it does not languish here...

Comment (kriegfrj) [reply bndtools.core/src/bndtools/central/Central.java:243]: Done; see #4582.

**Comment by kriegfrj** [comment]
> Looking at the failure, it seems unrelated to the previous. It seems that BuildPathQuickFixProcessorTest_WithSimpleOnClasspath failed to add Bndtools.core.test.fodder.simple to the -buildpath. I've never seen this before and when I reran it it passed.
> 
> Perhaps when it attempted to install the simple bundle onto -buildpath, the local repository hadn't finished updating yet?

I revisited this stray failure - I noted that the bnd.bnd file is updated direct and not through the Eclipse resources API, meaning that it is possible Eclipse may not have noticed. Added an explicit call to `Central.refresh()` in order to try and prevent this.

**Comment by bjhargrave** [comment]
@kriegfrj You feeling this is ready? If so, please remove the fail-fast commit and rebase.

**Comment by kriegfrj** [comment]
> @kriegfrj You feeling this is ready? If so, please remove the fail-fast commit and rebase.

I've learned by painful experience not to trust my feelings on this one... but let's give it a go...

I *think* this will solve #4253. I'm hopeful (but less confident) that it has solved #4553.

I think longer term I need to try and simplify the test a little bit to isolate the code-under-test from the other bndtools components. The test as it currently stands relies on other core Bndtools components like the builder and the cnf watcher. My original attempt at this was to build by calling Bnd's `Project.build()` method directly, but I couldn't get this to work. I now know that this was because I needed to refresh the build results. With more experience I might be able to do better. But I think what we have will work now, even if it is not 100% ideal.

**Comment by bjhargrave** [comment]
OK, its merged. Lets hope it remains stable :-)

**Comment by kriegfrj** [comment]
So far so good? I haven't seen any complaints?

**Comment by bjhargrave** [comment]
https://github.com/bndtools/bnd/runs/2399867371?check_suite_focus=true#step:5:4479 seems to be a random failure

**Comment by bjhargrave** [comment]
And another https://github.com/bndtools/bnd/runs/2404153924?check_suite_focus=true#step:5:4119

I think I want to disable this test in the CI build again.

**Comment by kriegfrj** [comment]
These two failure logs give me enough to go by to revisit and try and sort it out. This is unlikely to be soon so feel free to disable them in the meantime. The good thing is that the tests are still available to run any time someone is working on this code, which is when it is most important - having them run in CI would be icing on the cake.

The first failure seems to be a new failure mode. It is possible that Maven Central happened to be unreachable at the time and so it couldn't download the JUnit bundle. I could fix this by creating synthetic fodder classes that re-create the situation it is testing for and ditching the reliance on the external bundle.

The second seems like a variant of the original #4253. I now know that "Local Index" being null means that the workspace object hasn't been properly initialised. I'll have to have a closer look and see how this might be able to happen. I'll put my findings in the issue ticket.

**Comment by bjhargrave** [comment]
> It is possible that Maven Central happened to be unreachable

This happens more often than I like but I think we cannot be trying to work around this. I would not bother to do anything special here since some other place in the build will probably also fail.
